packages/appsync-modelgen-plugin/src/__tests__/utils/process-has-many.test.ts

```diff
@@ -0,0 +1,382 @@
+      isNullable: true,
+      directives: [],
+    };
+    const barModel: CodeGenModel = {
+      name: 'Bar',
+      type: 'model',
+      directives: [
+        { name: 'model', arguments: {} },
+      ],
+      fields: [
+        {
+          name: 'id',
+          type: 'ID',
+          isList: false,
+          isNullable: false,
+          directives: [],
+        },
+        associationField,
+      ],
+    }
+    const connection: CodeGenFieldConnectionHasMany = {
+      kind: CodeGenConnectionType.HAS_MANY,
+      connectedModel: barModel,
+      isConnectingFieldAutoCreated: true,
+      associatedWith: associationField,
+      associatedWithFields: [associationField],
+    };
+    expect(hasManyHasImplicitKey(hasManyField, fooModel, connection)).toBeTruthy();
+  });
+
+  it('returns false for implicit primary key with belongsTo', () => {
+    /**
+     * Example using the following schema:
+     *
+     * type Foo @model {
+     *   id: ID!
+     *   bar: [Bar] @hasMany
+     * }
+     *
+     * type Bar @model {
+     *   id: ID!
+     *   foo: Foo @belongsTo
+     * }
+     */
+    const hasManyField: CodeGenField = {
+      name: 'bar',
+      type: 'Bar',
+      isList: true,
+      isNullable: true,
+      directives: [{ name: 'hasMany', arguments: {} }],
+    }
+    const fooModel: CodeGenModel = {
+      name: 'Foo',
+      type: 'model',
+      directives: [
+        { name: 'model', arguments: {} },
+      ],
+      fields: [
+        {
+          name: 'id',
+          type: 'ID',
+          isList: false,
+          isNullable: false,
+          directives: [],
+        },
+        hasManyField,
+      ],
+    };
+    const associationField: CodeGenField = {
+      name: 'fooBarsId',
+      type: 'ID',
+      isList: false,
+      isNullable: true,
+      directives: [],
+    };
+    const barModel: CodeGenModel = {
+      name: 'Bar',
+      type: 'model',
+      directives: [
+        { name: 'model', arguments: {} },
+      ],
+      fields: [
+        {
+          name: 'id',
+          type: 'ID',
+          isList: false,
+          isNullable: false,
+          directives: [],
+        },
+        {
+          name: 'foo',
+          type: 'Foo',
+          isList: false,
+          isNullable: true,
+          directives: [{ name: 'belongsTo', arguments: {} }],
+        },
+        associationField,
+      ],
+    }
+    const connection: CodeGenFieldConnectionHasMany = {
+      kind: CodeGenConnectionType.HAS_MANY,
+      connectedModel: barModel,
+      isConnectingFieldAutoCreated: true,
+      associatedWith: associationField,
+      associatedWithFields: [associationField],
+    };
+    expect(hasManyHasImplicitKey(hasManyField, fooModel, connection)).toBeFalsy();
+  });
+
+  it('returns false for explicit index and id', () => {
+    /**
+     * Example using the following schema:
+     *
+     * type Foo @model {
+     *   id: ID!
+     *   bar: [Bar] @hasMany(indexName: "byFoo", fields: ["id"])
```

Comment:
what happens in case we have only `fields` but no `indexName` argument? It should still add the `key` right?

Reply:
It's not a valid use-case. We throw an InvalidDirectiveError in that case, since it's registered as a Partial sort key (which isn't exactly right either, but it's correctly erroring).